-
  In order to test the cancel sale order.
-
  I try to cancel sale order and It's order policy is Manual.
-
  !python {model: sale.order}: |
    try:
      self.action_cancel(cr, uid, [ref("order4")])
    except Exception,e:
      pass
-
  To cancel the sale order from Invoice Exception, I have to cancel the invoice of sale order.
-
  !python {model: sale.order}: |
    import netsvc
    wf_service = netsvc.LocalService("workflow")
    wf_service.trg_validate(uid, 'sale.order', ref('order3'), 'order_confirm', cr)

    invoice_ids = [x.id for x in self.browse(cr, uid, ref("order3")).invoice_ids]
    for invoice in invoice_ids:
      wf_service.trg_validate(uid, 'account.invoice', invoice, 'invoice_cancel', cr)

#code introduced by revid: mtr@tinyerp.com-20110915060653-cbedeckll1ijre66 
#form 'lp:~openerp-dev/openobject-addons/trunk-sale_coverage-mtr' branch

-
  I check order status in "Invoice Exception" and related invoice is in cancel state.
-
  !python {model: sale.order}: |
    order = self.browse(cr, uid, ref("order3"))
    assert order.invoice_ids[0].state == "cancel","order's related invoice should be cancelled"
    assert order.state == "invoice_except", "Order should be in Invoice Exception state after cancel Invoice"
-
  Then I click on the Ignore Exception button.
-
  !workflow {model: sale.order, action: invoice_corrected, ref: order3}
-
  And then again I set the invoice to draft state.
-
  !python {model: account.invoice}: |
    sale_order_obj = self.pool.get('sale.order')
    so = sale_order_obj.browse(cr, uid, ref("order3"))
    invoice_id = self.search(cr, uid, [('origin','=',so.name),('state','=','cancel')])
    self.action_cancel_draft(cr, uid, invoice_id, ({'active_model': 'ir.ui.menu','active_ids': [ref("sale.menu_sale_order")], 'type': 'out_invoice', 'active_id': ref("sale.menu_sale_order")},))
-
  Now I creating a partial picking.
-
  !python {model: sale.order}: |
    import netsvc
    order = self.browse(cr, uid, ref("order4"))
    assert order.picking_ids, "Picking is not created for this sale order"
    partial_pick = self.pool.get('stock.partial.picking')
    stock = self.pool.get('stock.picking')
    pick_ids = [x.id for x in self.browse(cr, uid, ref("order4")).picking_ids]
    stock1 = stock.browse(cr, uid, pick_ids)
    data = stock.force_assign(cr, uid, pick_ids)
    if data == True:
      partial_id = partial_pick.create(cr, uid, {},context={'active_model': 'stock.picking','active_ids': pick_ids})
      partial = partial_pick.browse(cr,uid,partial_id)
      line_id = partial.move_ids[0].id
      partial.write({'move_ids': [(1,line_id,{'quantity': 5})]})
      partial.do_partial()
      picking_id = stock.search(cr, uid, [('origin','=',order.name),('state','=','assigned')])
      wf_service = netsvc.LocalService("workflow")
      for picking in picking_id:
        wf_service.trg_validate(uid, 'stock.picking', picking, 'button_cancel', cr)

#code introduced by revid: mtr@tinyerp.com-20110921101038-xf1l3whyblmjbdx9
#form 'lp:~openerp-dev/openobject-addons/trunk-sale_coverage-mtr' branch
-
  Then I click on the "Recreate Packing" button on sale order
-
  !workflow {model: sale.order, action: ship_recreate, ref: order4}
-
  I verfiy that sale order state is now 'In Progress'.
-
  !assert {model: sale.order, id: order4}:
    - state == 'progress'
-
  I verify that picking is generated for sale order.
-
  !python {model: stock.picking}: |
    order = self.browse(cr, uid, ref("order4"))
    picking_id = self.search(cr, uid, [('sale_id','=',ref("order4"))])
    assert picking_id,"Picking has not been generated"
-
  Now I need to cancel the rekated picking of sale order.
-
  !python {model: sale.order}: |
    import netsvc
    order = self.browse(cr, uid, ref("order4"))
    picking_ids = [x.id for x in self.browse(cr, uid, ref("order4")).picking_ids]
    wf_service = netsvc.LocalService("workflow")
    for picking in picking_ids:
      wf_service.trg_validate(uid, 'stock.picking', picking, 'button_cancel', cr)
-
  Now I am able to cancel this sale order.
-
  !python {model: sale.order}: |
    self.action_cancel(cr, uid, [ref("order4")])
-
  I check that sale order is cancelled.
-
  !assert {model: sale.order, id: order4}:
    - state == 'cancel'
-
  Again set cancelled order to draft.
-
  !python {model: sale.order}: |
    self.action_cancel_draft(cr, uid, [ref("order4")])

